cmd/osmosisd/cmd/airdrop.go

```diff
@@ -0,0 +1,304 @@
+
+			totalAmount, ok := sdk.NewIntFromString(args[2])
+			if !ok {
+				return fmt.Errorf("failed to parse totalAmount: %s", args[2])
+			}
+
+			genFile := config.GenesisFile()
+			appState, genDoc, err := genutiltypes.GenesisStateFromGenFile(genFile)
+			if err != nil {
+				return fmt.Errorf("failed to unmarshal genesis state: %w", err)
+			}
+
+			authGenState := authtypes.GetGenesisStateFromAppState(cdc, appState)
+
+			accs, err := authtypes.UnpackAccounts(authGenState.Accounts)
+			if err != nil {
+				return fmt.Errorf("failed to get accounts from any: %w", err)
+			}
+
+			jsonFile, err := os.Open(filepath)
+			if err != nil {
+				return err
+			}
+			defer jsonFile.Close()
+
+			byteValue, _ := ioutil.ReadAll(jsonFile)
+
+			var genStateV036 GenesisStateV036
+
+			setCosmosBech32Prefixes()
+			err = aminoCodec.UnmarshalJSON(byteValue, &genStateV036)
+			if err != nil {
+				return err
+			}
+
+			snapshot := []SnapshotFields{}
+			balanceIndexByAddress := make(map[string]int)
+			totalAtomBalance := sdk.NewInt(0)
+			for index, account := range genStateV036.AppState.Accounts {
+				totalAtomBalance = totalAtomBalance.Add(account.Coins.AmountOf(denom))
+				balanceIndexByAddress[account.Address.String()] = index
+				snapshot = append(snapshot, SnapshotFields{
+					AtomAddress: account.Address.String(),
+					AtomBalance: account.Coins.AmountOf(denom),
+					AtomPercent: sdk.NewDec(0),
+				})
+			}
+
+			for _, delegation := range genStateV036.AppState.Staking.Delegations {
+				address := delegation.DelegatorAddress
+				index, ok := balanceIndexByAddress[address.String()]
+				if !ok {
+					continue
+				}
+				sharesInt := delegation.Shares.RoundInt()
+				snapshot[index].AtomBalance = snapshot[index].AtomBalance.Add(sharesInt)
+				totalAtomBalance = totalAtomBalance.Add(sharesInt)
+			}
+
+			for index, asnapshot := range snapshot {
+				amt := asnapshot.AtomBalance
+				percent := big.NewInt(0).Div(amt.Mul(sdk.NewInt(1000000)).BigInt(), totalAtomBalance.BigInt())
+				snapshot[index].AtomPercent = sdk.NewDecFromBigIntWithPrec(percent, 4)
+			}
+
+			params.SetBech32Prefixes()
+
+			balances := []banktypes.Balance{}
+			for index, account := range genStateV036.AppState.Accounts {
+				// fmt.Println("Address: " + account.Address.String())
+				// fmt.Println("Amount: " + account.Coins.String())
+
+				// create concrete account type based on input parameters
+				var genAccount authtypes.GenesisAccount
+				baseAccount := authtypes.NewBaseAccount(account.Address, nil, 0, 0)
+				genAccount = baseAccount
+
+				if err := genAccount.Validate(); err != nil {
+					return fmt.Errorf("failed to validate new genesis account: %w", err)
+				}
+
+				// Add the new account to the set of genesis accounts and sanitize the
+				// accounts afterwards.
+				accs = append(accs, genAccount)
+				accs = authtypes.SanitizeGenesisAccounts(accs)
+
+				atomAmt := account.Coins.AmountOf(denom)
+				osmoAmt, err := atomAmt.ToDec().ApproxSqrt()
```

Comment:
The square root of the unbonded balance and the bonded balances shouldn't be done separately.  It should be the sqrt of sum of the two.  Then, the multiplier should be done to percentage of the original balance that was staked.

So for example, if ATOM address has 40 staked and 10 unstaked, their OSMO balance would be:

sqrt (40 + 10)  +  sqrt (40 + 10) * 0.8  * 0.5

Reply:
Or separated out:

sqrt(bonded_atom + unbonded_atom)  *  (1 + (bonded_atom / (bonded_atom + unbonded_atom)) * 0.5)